flink-core/src/main/java/org/apache/flink/core/fs/SafeFileSystemFactory.java

```diff
@@ -18,29 +18,52 @@
 package org.apache.flink.core.fs;
 
 import org.apache.flink.configuration.Configuration;
+import org.apache.flink.util.FatalExitExceptionHandler;
 import org.apache.flink.util.TemporaryClassLoaderContext;
 import org.apache.flink.util.WrappingProxy;
+import org.apache.flink.util.concurrent.ExecutorThreadFactory;
 
 import java.io.IOException;
 import java.net.URI;
+import java.util.concurrent.CompletableFuture;
+import java.util.concurrent.CompletionException;
+import java.util.concurrent.ExecutorService;
+import java.util.concurrent.Executors;
 
 /**
- * A wrapper around {@link FileSystemFactory} that ensures the plugin classloader is used for all
- * {@link FileSystem} operations.
+ * A wrapper around {@link FileSystemFactory} that ensures the same classloader is used for all
+ * {@link FileSystem} operations and that we don't leak classloader via inherited {@link
+ * java.security.ProtectionDomain protection domains} when new threads that might outlive the
+ * classloader are spawned during {@link FileSystem} initialization.
  */
-public class PluginFileSystemFactory implements FileSystemFactory {
+public class SafeFileSystemFactory implements FileSystemFactory {
+
+    private static final ExecutorService EXECUTOR =
```

Comment:
We use the `ExecutorService` here (and instantiate a `Thread` right-away) so that the thread is created outside of the user-code context? This way, we ensure that the `FileSystem` is not loaded in user-code context. Does my reasoning match your intention?

Reply:
This is actually the main point of this change.

First off, we need to understand a bit more about how the classloading works.

Every class that is loaded is assigned a protection domain. The protection domain contains reference to a CodeSource ("class location, ie. path to jar" + signing certificates), reference to a ClassLoader that has loaded it, principals and  permissions.

The part that bothers us the most is the reference to the ClassLoader that has loaded the class. There is no way to get rid of it, that wouldn't involve re-implementing a good part of the low level ClassLoaders (because everything is marked as final...).

---

Now to the actual problem. There is a thing called `AccessControlContext`, which can be obtained by calling `AccessController.getContext()`. This context contains list of current protection domains.

In pseudo code, the list of protection domain is obtained as  follows:

```
getCurrentStackTrace() | getStackTraceElementClass | Class#getProtectionDomain | uniq (as in bash)
```

When a new thread is spawned it inherits the `AccessControlContext` of the calller.

So... if we spawn any thread that outlives a job from a place that has been called by user code (there is anything loaded by user classloader on stack), we'll leak the reference to user classloader to this newly created thread.

This happens especially with hadoop filesystem, where the initialization of the filesystem spawn some "cleaner threads" that are attached to the JVM lifecycle.

In case of web-sumission, file system initialization may be actually called from the user code (see the stacktrace - https://gist.github.com/dmvk/a77207634cde8094baf456430d1bd92d).

---

The way to work around this, is to spawn initialize filesystems in a "safe" thread, that has no potentially leaky protection domains, that the new thread could inherit.

This is unfortunately the cleanest solution I was able to come up with after 3 days of research.

Reply:
What you're describing makes sense. I struggled to find documentation about where the `AccessControlContext` is coming from when creating a new `Thread`.

Reply:
I'll add comments into the code.